### PR TITLE
Fixes potential race in JSON parser when parsing JSON lines format and when recovering from invalid lines

### PR DESCRIPTION
**Pull request by elstehle**
## Description

PR adds a missing synchronization before the FST destructor of the FST used for cleaning excess characters following the first valid record on a JSON line.

The problem is that the FST's destructor could otherwise free memory that is yet to be used by the still running FST instance.

Closes https://github.com/rapidsai/cudf/issues/15409

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/tests/io/json_test.cpp [hunk 1]
```diff
@@ -2050,6 +2054,110 @@ TEST_F(JsonReaderTest, JSONLinesRecoveringIgnoreExcessChars)
...
+{
+  // Set up host pinned memory pool to avoid implicit synchronizations to test for any potential
+  // races due to missing host-device synchronizations
+  using host_pooled_mr = rmm::mr::pool_memory_resource<rmm::mr::pinned_host_memory_resource>;
+  host_pooled_mr mr{std::make_shared<rmm::mr::pinned_host_memory_resource>().get(),
+                    size_t{128} * 1024 * 1024};
+
+  // Set new resource
+  auto last_mr = cudf::io::get_host_memory_resource();
+  cudf::io::set_host_memory_resource(mr);
```
Comment (abellina) [inline cpp/tests/io/json_test.cpp:2068]: `set_host_memory_resource` returns the prior `mr` so you could remove the call to `get_host_memory_resource` here.

**Review by abellina — APPROVED** [review]

**Comment by abellina** [comment]
Thanks for taking a look @elstehle

File: cpp/tests/io/json_test.cpp (hunk 1 above)
Comment (elstehle) [reply cpp/tests/io/json_test.cpp]: Great catch, thanks 👍

**Review by shrshi — APPROVED** [review]
Looks good to me! :)

File: cpp/src/io/json/nested_json_gpu.cu [hunk 2]
```diff
@@ -1583,6 +1583,9 @@ std::pair<rmm::device_uvector<PdaTokenT>, rmm::device_uvector<SymbolOffsetT>> ge
                                         thrust::make_discard_iterator(),
                                         fix_stack_of_excess_chars::start_state,
                                         stream);
+
+    // Make sure memory of the FST's lookup tables isn't freed before the FST completes
```
Comment (vuule) [inline cpp/src/io/json/nested_json_gpu.cu:1587]: I assume this is about `fix_stack_of_excess_chars`. Which part of it is not destroyed in stream order (AFAICT, must be a host-side buffer?)?

Comment (vuule) [reply cpp/src/io/json/nested_json_gpu.cu:1587]: @shrshi answered this offline :)

**Review by vuule — APPROVED** [review]

**Pull request merged**